File: framework/languages/java/e2e-kotlin/ToActorMessaging/Shared/src/main/java/systems/zlink/e2e/kotlin/toactormessaging/shared/Env.java
package systems.zlink.e2e.kotlin.toactormessaging.shared;

import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Properties;

public final class Env {
    private static Properties values;

    private Env() {
    }

    public static void configure(String[] args) {
        if (args.length != 2 || !"--config".equals(args[0]) || args[1].isBlank()) {
            throw new IllegalArgumentException("Usage: <role> --config <path>");
        }
        Properties loaded = new Properties();
        try (Reader reader = Files.newBufferedReader(Path.of(args[1]))) {
            loaded.load(reader);
        } catch (Exception error) {
            throw new IllegalStateException("Could not load Kotlin ToActorMessaging config.", error);
        }
        values = loaded;
    }

    public static String get(String key) {
        String value = properties().getProperty(key);
        if (value == null || value.isBlank()) {
            throw new IllegalStateException("Missing configuration value " + key);
        }
        return value;
    }

    public static String get(String key, String fallback) {
        String value = properties().getProperty(key);
        return value == null || value.isBlank() ? fallback : value;
    }

    private static Properties properties() {
        if (values == null) {
            throw new IllegalStateException("Kotlin ToActorMessaging configuration was not loaded.");
        }
        return values;
    }
}

File: framework/languages/java/e2e-kotlin/ResilienceLifecycle/Shared/src/main/kotlin/systems/zlink/e2e/kotlin/resiliencelifecycle/Env.kt
package systems.zlink.e2e.kotlin.resiliencelifecycle

import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties

object Env {
    private val values = linkedMapOf<String, String>()

    @JvmStatic
    @Synchronized
    fun configure(args: Array<out String>) {
        values.clear()
        val path = configPath(args) ?: return
        val properties = Properties()
        Files.newInputStream(Path.of(path)).use { properties.load(it) }
        for (name in properties.stringPropertyNames()) {
            values[name] = properties.getProperty(name)
        }
    }

    @JvmStatic
    fun get(name: String): String =
        values[name] ?: throw IllegalStateException("missing config value $name")

    @JvmStatic
    fun get(name: String, defaultValue: String): String =
        values[name] ?: defaultValue

    @JvmStatic
    fun applicationArgs(args: Array<out String>): Array<String> = withoutConfig(args)

    private fun configPath(args: Array<out String>): String? {
        for (index in args.indices) {
            val argument = args[index]
            if (argument == "--e2e-config" && index + 1 < args.size) {
                return args[index + 1]
            }
            if (argument.startsWith("--e2e-config=")) {
                return argument.substringAfter('=', "")
            }
        }
        return null
    }

    private fun withoutConfig(args: Array<out String>): Array<String> {
        val result = mutableListOf<String>()
        var index = 0
        while (index < args.size) {
            val argument = args[index]
            if (argument == "--e2e-config") {
                index += 2
            } else if (argument.startsWith("--e2e-config=")) {
                index++
            } else {
                result += argument
                index++
            }
        }
        return result.toTypedArray()
    }

}
